doc/adrs/XXX_add_support_for_backend-driven_diagram_layout_customization.adoc

```diff
@@ -0,0 +1,185 @@
+= ADR-XXX - Add support for backend-driven diagram customization
+
+== Context
+
+The current Sirius Web architecture handles diagram updates via a mix of frontend and backend logic tightly coupled to layout and rendering concerns.
+No APIs are available to programmatically influence diagram content or behavior.
+
+This architectural rigidity prevents advanced use cases such as automated transformations, semantic-aware or rule-based diagram evolution.
+
+== Needs
+
+We need to:
+
+- Support structural changes like adding, removing, or modifying diagram elements.
+- Allow injection of domain-specific logic while preserving compatibility with existing mechanisms (e.g., rendering, persistence, sync).
+- Support composability and fallback to default behavior when not overridden.
+
+Such extensibility is currently impossible due to tight coupling between the backend/frontend pipeline and layout-specific logic.
+
+== Alternatives
+
+=== [Frontend] Frontend-side customization
+- Would require loading full diagram and semantic context client-side.
+- Increased complexity and negative performance impact.
+- Limited control over backend-managed logic.
+
+=== [Backend] Backend-side customization
+- Direct access to diagram and semantic data.
+- Easier integration with rendering and persistence layers.
+- Better performance and extensibility.
+
+== Decision
+
+We introduce an API mechanism in `DiagramCreationService` to allow post-processing of diagrams after they are created or updated.
+This is done through a new interface, `IDiagramPostProcessor`, that lets us plug in custom logic easily.
+
+The goal is to make it possible to adjust a diagram (layout and/or semantic changes, etc.) without modifying the core computation.
+
+This approach is inspired by a similar mechanism already used for forms (`IFormPostProcessor`), and helps keep things simple and consistent.
+
+
+
+=== API for custom diagram logic Providers
+
+A new backend API is introduced to support domain-specific or diagram-type-specific strategies.
+Furthermore, the specifier can define which specific actions (`IDiagramInput` interface) should trigger these custom logics, such as `EditLabelInput`, `DeleteFromDiagramInput`, etc.
+This brings more flexibility to specifiers to adapt diagram/layout behavior to the needs of their diagrams or domain logic.
+
+==== IDiagramPostProcessor interface
+
+[source,java]
+----
+public interface IDiagramPostProcessor {
+
+    /**
+     * Applies custom logic to a diagram after creation or update.
+     *
+     * @param diagram The diagram to process
+     * @param diagramDescription Optional diagram description (if already available)
+     * @param diagramContext Optional context containing triggering information
+     * @return The modified diagram
+     */
+    Diagram postProcess(Diagram diagram, Optional<DiagramDescription> diagramDescription, Optional<IDiagramContext> diagramContext);
```

Comment:
You will need the `IEditingContext` too to make it project specific for downstream consumers